File: leetcode/157.read-n-characters-given-read4.cpp
// Forward declaration of the read4 API.
int read4(char *buf);

class Solution {
public:
    /**
     * @param buf Destination buffer
     * @param n   Maximum number of characters to read
     * @return    The number of characters read
     */
    int read(char *buf, int n) {
        char buffer[4];
        int offset = 0;
        int total = 0;

        for(auto i = 0; i < n; ++i)
        {
            if (offset == total)
            {
                total = read4(buffer);
                offset = 0;
            }

            if (offset < total){
                buf[i] = buffer[offset++];
            }
            else
            {
                return i;
            }
        }

        return n;
    }
};

File: leetcode/157.read-n-characters-given-read4.py
# The read4 API is already defined for you.
# @param buf, a list of characters
# @return an integer
# def read4(buf):

class Solution(object):
    def read(self, buf, n):
        """
        :type buf: Destination buffer (List[str])
        :type n: Maximum number of characters to read (int)
        :rtype: The number of characters read (int)
        """
        buffer = [0] * 4
        offset = 0
        total = 0

        for i in range(n):
            if offset == total:
                total = read4(buffer)
                offset = 0

            if offset < total:
                buf[i] = buffer[offset]
                offset += 1
            else:
                return i

        return n